lib/galaxy/tool_util/linters/inputs.py

```diff
@@ -469,6 +544,93 @@ def lint_inputs(tool_xml, lint_ctx):
+        if _param_in_compiled_cheetah(param, param_name, code):
+            continue
+        if param_name in filter_code:
+            continue
+        if (
+            tool_xml.find("./inputs//param/options[@from_dataset]") is not None
+            or tool_xml.find("./inputs//param/options/filter[@ref]") is not None
+        ):
+            continue
+        if param.getparent().tag == "conditional":
+            continue
+
+        conf_inputs = tool_xml.find("./configfiles/inputs")
+        if conf_inputs is not None:  # in this it's really hard to know
+            param_type = param.attrib.get("type")
+            if param_type in ["data", "collection"]:
+                if not conf_inputs.attrib.get("data_style"):
+                    lint_ctx.error(
+                        f"Param input [{param_name}] not found in command or configfiles. Does the present inputs config miss the 'data_style' attribute?",
+                        node=param,
+                    )
+            inputs_conf_name = conf_inputs.attrib.get("name", conf_inputs.attrib.get("filename", None))
+            if inputs_conf_name:
+                if not re.search(r"(^|[^\w])" + inputs_conf_name + r"([^\w]|$)", code):
+                    lint_ctx.error(
+                        f"Param input [{param_name}] only used inputs configfile {inputs_conf_name}, but this is not used in the command",
+                        node=param,
+                    )
+            continue
+
+        change_format = tool_xml.find(f"./outputs//change_format/when[@input='{param_name}']") is not None
+        template = re.search(r"(^|[^\w])" + param_name + r"([^\w]|$)", template_code) is not None
+        action = re.search(r"(^|[^\w])" + param_name + r"([^\w]|$)", action_code) is not None
+        label = re.search(r"[^\w]" + param_name + r"([^\w]|$)", label_code) is not None
+        if template or action or label:
+            if template + action + label == 1:
+                only = "only "
+            else:
+                only = ""
+            # TODO check that template is used??
+            if template:
+                lint_ctx.warn(
+                    f"Param input [{param_name}] {only}used in a template macro, use a macro token instead.", node=param
+                )
+            if action:
+                lint_ctx.warn(
+                    f"Param input [{param_name}] {only}found in output actions, better use extended metadata.",
+                    node=param,
+                )
+            if label:
+                lint_ctx.warn(
+                    f"Param input [{param_name}] {only}found in a label attribute, this is discouraged.", node=param
+                )
+            continue
+
+        if change_format:
+            lint_ctx.error(f"Param input [{param_name}] is only used in a change_format tag", node=param)
+        else:
+            lint_ctx.error(f"Param input [{param_name}] not found in command or configfiles.", node=param)
```

Comment:
error should only be used if the tool is broken with that change IMO. I don't think unused variables really indicate a tool is broken. I would make these warns or infos.

Reply:
In particular the last case (the `else` branch) is clearly an error in my opinion. The parameter is used nowhere in the tool, i.e. the functionality of this parameter is broken (even if the tool still runs...). Just assume it would be the light switch in a car .. if it does not switch the lights on/off it's broken even if the car is still driving, or?

For the `if` branch I just can not imagine how the tool may produce a different format without using the parameter anywhere else except for the `change_format` tag.

But I'm fine with downgrading errors and warnings to warnings and infos (resp) if its for the possibility of FP due to `getVar` and `varExists`...

Reply:
For `getVar` and `varExists` we could just bail out like for the case when a param file is present